File: Algorithm_104_Maximum Depth of Binary Tree.cpp
#include <iostream>
using namespace std;
int maxLen=0;
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
int dep(TreeNode *root,int thisLen)
{
	if (root!=NULL)
	{
		thisLen++;
		if (thisLen>maxLen) maxLen=thisLen;
		dep(root->left,thisLen);
		dep(root->right,thisLen);
	}
	return maxLen;
}
int maxDepth(TreeNode* root)
{
	return dep(root,0);
}
int main()
{
	TreeNode *t1=new TreeNode(1);
	TreeNode *t2=new TreeNode(2);
	TreeNode *t3=new TreeNode(3);
	TreeNode *t4=new TreeNode(4);
	t1->left=t2;
	t1->right=t3;
	t2->left=t4;
	cout<<maxDepth(t1);
}

File: Algorithm_104_Maximum Depth of Binary Tree.java
// 思路：直接对左右子节点递归

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int maxDepth(TreeNode root) {
        return root == null ? 0 : 1 + Math.max(maxDepth(root.left), maxDepth(root.right));
    }
}

// BFS Solution
// int maxDepth(TreeNode *root)
// {
//     if(root == NULL)
//         return 0;
    
//     int res = 0;
//     queue<TreeNode *> q;
//     q.push(root);
//     while(!q.empty())
//     {
//         ++ res;
//         for(int i = 0, n = q.size(); i < n; ++ i)
//         {
//             TreeNode *p = q.front();
//             q.pop();
            
//             if(p -> left != NULL)
//                 q.push(p -> left);
//             if(p -> right != NULL)
//                 q.push(p -> right);
//         }
//     }
    
//     return res;
// }